packages/form-core/src/util-types.ts

```diff
@@ -83,17 +85,40 @@ type PrefixFromDepth<
   TDepth extends any[],
 > = TDepth['length'] extends 0 ? T : `.${T}`
 
-export type DeepValue<TValue, TAccessor> = TValue extends Record<
-  string | number,
-  any
->
-  ? TAccessor extends `${infer TBefore}[${infer TBrackets}].${infer TAfter}`
-    ? DeepValue<TValue[TBefore][TBrackets], TAfter>
-    : TAccessor extends `[${infer TBrackets}]`
-      ? DeepValue<TValue, TBrackets>
-      : TAccessor extends `${infer TBefore}[${infer TBrackets}]`
-        ? DeepValue<TValue[TBefore], TBrackets>
-        : TAccessor extends `${infer TBefore}.${infer TAfter}`
-          ? DeepValue<TValue[TBefore], TAfter>
-          : TValue[TAccessor & string]
-  : never
+export type DeepValue<
+  TValue,
+  TAccessor,
+  TDepth extends any[] = [],
+> = TDepth['length'] extends 3
+  ? never
+  : TValue extends ReadonlyArray<any>
+    ? TAccessor extends `[${infer TBrackets}].${infer TAfter}`
+      ? DeepValue<
+          DeepValue<TValue, TBrackets, [...TDepth, any]>,
+          TAfter,
+          [...TDepth, any]
+        >
+      : TAccessor extends `[${infer TBrackets}]`
+        ? DeepValue<TValue, TBrackets, [...TDepth, any]>
+        : TAccessor extends keyof TValue
+          ? TValue[TAccessor]
+          : TValue[TAccessor & number]
+    : TValue extends Record<string | number, any>
+      ? TAccessor extends `${infer TBefore}[${infer TEverythingElse}`
+        ? DeepValue<
+            DeepValue<TValue, TBefore, [...TDepth, any]>,
+            `[${TEverythingElse}`,
+            [...TDepth, any]
+          >
+        : TAccessor extends `[${infer TBrackets}]`
+          ? DeepValue<TValue, TBrackets, [...TDepth, any]>
+          : TAccessor extends `${infer TBefore}.${infer TAfter}`
+            ? DeepValue<
+                DeepValue<TValue, TBefore, [...TDepth, any]>,
+                TAfter,
+                [...TDepth, any]
+              >
+            : TAccessor extends string
+              ? TValue[TAccessor]
+              : never
+      : never
```

Comment:
Okay, these are the comments I'd add to this type. 

```suggestion
/**
   Infer the type of a deeply nested property within an object or an array.
   See the tests in `util-types.test-d.ts` for usage.
*/
export type DeepValue<
  // The object or array in which we have the property whose type we're trying to infer
  TValue,
  // A string representing the path of the property we're trying to access
  TAccessor,
  // (Optional) An array representing the current depth of recursion
  TDepth extends any[] = [],
> =
  // Limit the depth of recursion to 3 levels
  TDepth['length'] extends 3
    ? never
    : // Check if we're looking for the property in an array
      TValue extends ReadonlyArray<any>
      ? TAccessor extends `[${infer TBrackets}].${infer TAfter}`
        ? /*
          Extract the first element from the accessor path (`TBrackets`)
          and recursively call `DeepValue` with it
          */
          DeepValue<
            DeepValue<TValue, TBrackets, [...TDepth, any]>,
            TAfter,
            [...TDepth, any]
          >
        : TAccessor extends `[${infer TBrackets}]`
          ? DeepValue<TValue, TBrackets, [...TDepth, any]>
          : TAccessor extends keyof TValue
            ? TValue[TAccessor]
            : TValue[TAccessor & number]
      : // Check if we're looking for the property in an object
        TValue extends Record<string | number, any>
        ? TAccessor extends `${infer TBefore}[${infer TEverythingElse}`
          ? DeepValue<
              DeepValue<TValue, TBefore, [...TDepth, any]>,
              `[${TEverythingElse}`,
              [...TDepth, any]
            >
          : TAccessor extends `[${infer TBrackets}]`
            ? DeepValue<TValue, TBrackets, [...TDepth, any]>
            : TAccessor extends `${infer TBefore}.${infer TAfter}`
              ? DeepValue<
                  DeepValue<TValue, TBefore, [...TDepth, any]>,
                  TAfter,
                  [...TDepth, any]
                >
              : TAccessor extends string
                ? TValue[TAccessor]
                : never
        : // Do not allow `TValue` to be anything else
          never

```

Reply:
Love what I'm seeing. Let's move this to a nested PR (a PR onto this PR) so we can discuss more thoroughly

Reply:
Awesome! Here you go: https://github.com/TanStack/form/pull/634